### PR TITLE
Use a taxonomy instead of meta for template & template-part 

### PR DESCRIPTION
**Pull request by aristath**
As discussed in #24377 and a [core editor meeting](https://wordpress.slack.com/archives/C02QB2JS7/p1597848789344200), we want to move away from using post-meta to store the theme, and start using a taxonomy instead.

This PR adds the taxonomy and changes all related queries to use the new `wp_theme` tax.

### CODE REVIEW

File: lib/templates.php [hunk 1]
```diff
@@ -55,6 +55,22 @@ function gutenberg_register_template_post_type() {
...
+ * Registers block editor 'wp_theme' taxonomy.
+ */
+function gutenberg_register_wp_theme_taxonomy() {
+
+	// Register wp_theme taxonomy.
+	register_taxonomy(
+		'wp_theme',
+		array( 'wp_template', 'wp_template_part' ),
+		array(
+			'public' => true,
```
[the remark this answers is not in the document]
Comment (aristath) [reply lib/templates.php]: taxonomy is now hidden

**Comment by aristath** [comment]
Terms now get properly created and assigned.
The queries seem to be working correctly.

**Review by tomjn — APPROVED** [review]

**Comment by aristath** [comment]
I currently have an issue with this one... Theme-defined template-parts don't show up in the site editor :thinking:

**Comment by aristath** [comment]
On the PHP front this is OK, now working on the JS front.
Since this was imcomplete (I previously missed all the `meta.theme` references in the template-part JS files) I converted this PR to a draft so we can keep working on it.

**Comment by TimothyBJacobs** [comment]
If this is using a taxonomy now, can we use the automated handling we'd get for querying and response inclusion by marking it as `show_in_rest`?

**Comment by aristath** [comment]
> If this is using a taxonomy now, can we use the automated handling we'd get for querying and response inclusion by marking it as `show_in_rest`?

Thank you @TimothyBJacobs! I added `show_in_rest` to the taxonomy now, but I'm afraid I don't quite understand what you mean by _"use the automated handling we'd get for querying and response inclusion"_, could you please elaborate?
Right now I'm exposing `wp_theme` in the JSON response using this: https://github.com/WordPress/gutenberg/pull/24720/files#diff-51f3ab0134282ae12e1994208a3ab6feR261, and then I use that instead of the meta we previously had. So `templatePart.meta.theme` just becomes `templatePart.wp_theme`. Are you suggesting there's a better way?

**Comment by TimothyBJacobs** [comment]
Now that the taxonomy shows in REST, you should be able to query for it [automatically](https://github.com/WordPress/wordpress-develop/blob/f60aa519882babdf934fda8f974c05fef84607ef/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2726) using the `rest_base` of the taxonomy. 

It should also now be automatically included in the [response](https://github.com/WordPress/wordpress-develop/blob/f60aa519882babdf934fda8f974c05fef84607ef/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2393) as the `rest_base` too.

**Comment by aristath** [comment]
@TimothyBJacobs 
Ah OK, I see what you mean now.
The problem with the default responses however is that they use the term-ID, and what we have here is the term-slug.
So I don't think we'd be able to get what we need without an additional REST call... Instead of doing that, I simply added the term-slug in the response.
Note: after your comment I realized that there was a conflict with default responses. By default it was returning `wp_theme: [ 2 ]` so I renamed the slug to `wp_theme_slug` and it doesn't overwrite the `wp_theme` anymore in responses.

If you feel this can be done better then please feel free to push commits here, you are far more familiar with the REST API than I am!

**Comment by aristath** [comment]
Superseded by https://github.com/WordPress/gutenberg/pull/27016
Closing this one since the other PR is up to date with the latest structural changes and it will be easier to work on.

**Pull request closed without merging**